### PR TITLE
Feature/use handle api for pid creation

### PR DESCRIPTION
**Pull request by southeo**
Implementing functionality to create handles using a webclient.
Have not implemented this in the specimen pipeline. Existing functionality is not affected in this PR.  

Moved some code around, so there are some bigger structural changes:

HandleComponent: This now contains a webclient function that connects with the handle API and posts a JSON. 

FdoRecordBuilder: Given a list of digital specimens, generate a JSON request body to send to the handle component. 

- This also  contains the code for what was the HandleService, i.e. build the list of HandleAttributes and post to the repository layer. That functionality will be removed in an upcoming PR, replaced by the HandleComponent communication with the Handle API.

### CODE REVIEW

File: src/test/java/eu/dissco/core/digitalspecimenprocessor/utils/TestUtils.java [hunk 1]
```diff
@@ -38,6 +38,7 @@ public class TestUtils {
   public static String SOURCE_SYSTEM_ID = "20.5000.1025/MN0-5XP-FFD";
   public static JsonNode ORIGINAL_DATA = generateSpecimenOriginalData();
   public static String DWCA_ID = null;
+  public static String TOPIC_DISCIPLINE = "geology";
```
Comment (samleeflang) [inline src/test/java/eu/dissco/core/digitalspecimenprocessor/utils/TestUtils.java:41]: Is this used? The topic discipline for geology is "earth systems" (geology is not a topic discipline)

**Review by samleeflang — COMMENTED** [review]
Some comments but looking great

File: src/main/java/eu/dissco/core/digitalspecimenprocessor/component/FdoRecordBuilder.java [hunk 2]
```diff
@@ -88,6 +72,75 @@ public class HandleService {
...
+    var attributes = mapper.createObjectNode();
+    // Defaults
+    attributes.put(FdoProfileAttributes.FDO_PROFILE.getAttribute(), FdoProfileConstants.FDO_PROFILE.getValue());
+    attributes.put(FdoProfileAttributes.DIGITAL_OBJECT_TYPE.getAttribute(), FdoProfileConstants.DIGITAL_OBJECT_TYPE.getValue());
+    attributes.put(FdoProfileAttributes.ISSUED_FOR_AGENT.getAttribute(), FdoProfileConstants.ISSUED_FOR_AGENT_PID.getValue());
+
+    // Mandatory
+    attributes.put(FdoProfileAttributes.PRIMARY_SPECIMEN_OBJECT_ID.getAttribute(), specimen.physicalSpecimenId());
+    var orgId = getTerm(specimen, "ods:organisationId");
+    orgId.ifPresent(s->attributes.put(SPECIMEN_HOST.getAttribute(), s));
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/digitalspecimenprocessor/component/FdoRecordBuilder.java:104]: I tried, but you can't throw an exception with ifPresentOrElse() :/ 

Made the lambda more clear though

File: src/main/java/eu/dissco/core/digitalspecimenprocessor/component/FdoRecordBuilder.java [hunk 3]
```diff
@@ -88,6 +72,75 @@ public class HandleService {
...
+    }
+  }
+
+  private void updateOptionalAttribute(DigitalSpecimen specimen, String term, String fdoAttribute, ObjectNode attributeNode){
+    var optionalAttribute = getTerm(specimen, term);
+    optionalAttribute.ifPresent(foundTerm -> attributeNode.put(fdoAttribute, foundTerm));
+  }
+
+  private Optional<String> getTerm(DigitalSpecimen specimen, String term){
+    var val = specimen.attributes().findValue(term);
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/digitalspecimenprocessor/component/FdoRecordBuilder.java:135]: changed

File: src/main/java/eu/dissco/core/digitalspecimenprocessor/configuration/HandleConnectorConfig.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.http.client.reactive.ReactorClientHttpConnector;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.netty.http.client.HttpClient;
+
+@Configuration
+public class HandleConnectorConfig {
+  @Bean
+  public WebClient handleResolverClient() {
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/core/digitalspecimenprocessor/configuration/HandleConnectorConfig.java:12]: Removed

File: src/main/java/eu/dissco/core/digitalspecimenprocessor/component/HandleComponent.java [hunk 5]
```diff
@@ -11,36 +15,60 @@
...
+            .filter(WebClientUtils::is5xxServerError)
+            .onRetryExhaustedThrow((retryBackoffSpec, retrySignal) ->
+                new PidCreationException(
+                    "External Service failed to process after max retries")
+            ));
+    try {
+      return response.toFuture().get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    } catch (ExecutionException e) {
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/digitalspecimenprocessor/component/HandleComponent.java:60]: I would do it slightly differently. Keep the HandleComponent only for the call. If it throws an exception let the ProcessingService handle those exceptions. That is the cooridinator of the whole flow.

File: src/main/java/eu/dissco/core/digitalspecimenprocessor/component/HandleComponent.java [hunk 6]
```diff
@@ -11,36 +15,60 @@
...
+      if (e.getCause().getClass().equals(PidAuthenticationException.class)) {
+        kafkaService.deadLetterEvent(digitalSpecimenEvents);
+        throw new PidAuthenticationException(e.getCause().getMessage());
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/digitalspecimenprocessor/component/HandleComponent.java:63]: Is this issue logged somewhere?

File: src/main/java/eu/dissco/core/digitalspecimenprocessor/component/TokenAuthenticator.java [hunk 7]
```diff
@@ -12,39 +13,57 @@
...
+            .onRetryExhaustedThrow((retryBackoffSpec, retrySignal) ->
+                new PidAuthenticationException(
+                    "External Service failed to process after max retries")
+            ));
+    try {
+      var tokenNode = response.toFuture().get();
+      return getToken(tokenNode);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      return null;
```
Comment (samleeflang) [inline src/main/java/eu/dissco/core/digitalspecimenprocessor/component/TokenAuthenticator.java:48]: Maybe log what has happened

**Review by samleeflang — COMMENTED** [review]
Some last comments. I think the handle service shouldn't be called from the HandleComponent. Let the HandleComponent throw an error and the processing service handle the failure. This also contains the other logic for handling failures. Otherwise nice changes for the retry and client exception handling. Also nice find on the testing :+1:

**Review by samleeflang — APPROVED** [review]
Thanks for changes, looks good! Let's get this through

**Pull request merged**